sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala

```diff
@@ -0,0 +1,338 @@
+      // This setting selects the original local timestamp mapped to the given `micros`.
+      .set(Calendar.DST_OFFSET, zoneId.getRules.getDaylightSavings(instant).toMillis.toInt)
+      .build()
+    millisToMicros(cal.getTimeInMillis) + ldt.get(ChronoField.MICRO_OF_SECOND)
+  }
+
+
+  /**
+   * The class describes JSON records with microseconds rebasing info.
+   * Here is an example of JSON file:
+   * {{{
+   *   [
+   *     {
+   *       "tz": "Europe/Paris",
+   *       "switches": [-123, 0],
+   *       "diffs": [422000000, 0]
+   *     }
+   *   ]
+   * }}}
+   *
+   * @param tz One of time zone ID which is expected to be acceptable by `ZoneId.of`.
+   * @param switches An ordered array of seconds since the epoch when the diff between
+   *                 two calendars are changed.
+   * @param diffs Differences in seconds associated with elements of `switches`.
+   */
+  private case class JsonRebaseRecord(tz: String, switches: Array[Long], diffs: Array[Long])
+
+  /**
+   * Rebasing info used to convert microseconds from an original to a target calendar.
+   *
+   * @param switches An ordered array of microseconds since the epoch when the diff between
+   *                 two calendars are changed.
+   * @param diffs Differences in microseconds associated with elements of `switches`.
+   */
+  private[sql] case class RebaseInfo(switches: Array[Long], diffs: Array[Long])
+
+  /**
+   * Rebases micros since the epoch from an original to an target calendar, for instance,
+   * from a hybrid (Julian + Gregorian) to Proleptic Gregorian calendar.
+   *
+   * It finds the latest switch micros which is less than `value`, and adds the difference
+   * in micros associated with the switch micros to the given `value`.
+   * The function is based on linear search which starts from the most recent switch micros.
+   * This allows to perform less comparisons for modern timestamps.
+   *
+   * @param rebaseInfo The rebasing info contains an ordered micros when difference in micros
+   *                   between original and target calendar was changed,
+   *                   and differences in micros between calendars
+   * @param micros The number of micros since the epoch 1970-01-01T00:00:00Z to be rebased
+   *               to the target calendar. It can be negative.
+   * @return The rebased micros.
+   */
+  private def rebaseMicros(rebaseInfo: RebaseInfo, micros: Long): Long = {
+    val switches = rebaseInfo.switches
+    var i = switches.length
+    do { i -= 1 } while (i > 0 && micros < switches(i))
+    micros + rebaseInfo.diffs(i)
+  }
+
+  // Loads rebasing info from an JSON file. JSON records in the files should conform to
+  // `JsonRebaseRecord`. AnyRefMap is used here instead of Scala's immutable map because
+  // it is 2 times faster in DateTimeRebaseBenchmark.
+  private[sql] def loadRebaseRecords(fileName: String): AnyRefMap[String, RebaseInfo] = {
```

Comment:
Let's try to don't `private[sql]` next time per SPARK-16813 next time.

Reply:
The reason "It simply makes debugging harder when Spark developers need to inspect the plans at runtime." is not applicable here, I believe. I would make the method `private` but I needed to access it from tests.

Reply:
Seems like it's the same package `org.apache.spark.sql.catalyst.util`. I guess `private[util]` works.